### PR TITLE
CNS-784: Projects README

### PR DESCRIPTION
**Pull request by oren-lava**

### CODE REVIEW

File: x/projects/README.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  * [Project Keys YAML](#project-keys-yaml)
+* [Proposals](#proposals)
+
+## Concepts
+
+### Project
+
+Once a consumer purchases a subscription, they can create projects under it. Each project has a unique name (index) within this subscription. A project is defined as follows:
+
+```
```
Comment (shleikes) [inline x/projects/README.md:30]: Add the `go` to highlight syntax 

```suggestion
```go
```

File: x/projects/README.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+Projects are saved in a fixation store (see `FixationStore` module for reference). By using the snapshot ID, we can differentiate between different versions of the project.
+
+For more details regarding the `Policy` struct and its limitations, see [here](https://github.com/lavanet/lava/blob/main/x/plans/README.md#policy). For more details about `ProjectKeys`, see [below](#project-keys).
+
+### Project Keys
+
+A project has a list of accounts (project keys) that can use the project in different ways. There are two kinds of keys: admin key and developer key. A project key is defined as follows:
+
+```
```
Comment (shleikes) [inline x/projects/README.md:55]: Add the `go` to highlight syntax 

```suggestion
```go
```

File: x/projects/README.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+* [Proposals](#proposals)
+
+## Concepts
+
+### Project
+
+Once a consumer purchases a subscription, they can create projects under it. Each project has a unique name (index) within this subscription. A project is defined as follows:
+
+```
+struct Project {
```
Comment (shleikes) [inline x/projects/README.md:31]: Why no:

```suggestion
type Project struct {
```

File: x/projects/README.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+Projects are saved in a fixation store (see `FixationStore` module for reference). By using the snapshot ID, we can differentiate between different versions of the project.
+
+For more details regarding the `Policy` struct and its limitations, see [here](https://github.com/lavanet/lava/blob/main/x/plans/README.md#policy). For more details about `ProjectKeys`, see [below](#project-keys).
+
+### Project Keys
+
+A project has a list of accounts (project keys) that can use the project in different ways. There are two kinds of keys: admin key and developer key. A project key is defined as follows:
+
+```
+struct ProjectKey {
```
Comment (shleikes) [inline x/projects/README.md:56]: Why not:

```suggestion
type ProjectKey struct {
```

File: x/projects/README.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+Note that the admin cannot use the project's CU like a developer, they can only edit the project's properties.
+
+### Badges
+
+Badges are a method to grant compute units (CU) from projects to other users. Given a subscription with projects, a developer key (in a project) can be used to generate a badge key, to be signed by an external badge-server. The badge is ephemeral, limited in time and in CU capacity; It can be used to get pairing with providers, and for requests from providers - which will be charged to the respective project after verification.
+
+A badge is defined as follows:
+
+```
+struct Badge {
```
Comment (shleikes) [inline x/projects/README.md:85]: Same as previous comment

File: x/plans/README.md [hunk 6]
```diff
@@ -113,7 +113,7 @@ The `InternalPath` field is utilized for chains that have varying RPC API sets i
 
 The `Type` field lets the user define APIs that have different functionalities depending on their type. the valid types are: `GET` and `POST`. An example of such API is Cosmos' `/cosmos/tx/v1beta1/txs` API. If it's sent as a `GET` request, it fetches transactions by event and if it's sent as a `POST` request, it sends a transaction.
 
-The `AddOn` field lets you use additional optional APIs like debug, trace, 
+The `AddOn` field lets you use additional optional APIs (debug, trace and more). Overall, the add-ons can be: `debug`, `trace` and `convex`.
```
Comment (shleikes) [inline x/plans/README.md:116]:
```suggestion
The `AddOn` field lets you use additional optional APIs, which can be: `debug`, `trace` and `convex`.
```

**Review by shleikes — CHANGES_REQUESTED** [review]

File: x/projects/README.md (hunk 1 above)
Comment (oren-lava) [reply x/projects/README.md]: done. see https://github.com/lavanet/lava/commit/576717fd44c05504a0127dc80eb39d09a153b012

File: x/projects/README.md (hunk 2 above)
Comment (oren-lava) [reply x/projects/README.md]: done. see https://github.com/lavanet/lava/commit/576717fd44c05504a0127dc80eb39d09a153b012

File: x/projects/README.md (hunk 3 above)
Comment (oren-lava) [reply x/projects/README.md]: done. see https://github.com/lavanet/lava/commit/576717fd44c05504a0127dc80eb39d09a153b012

File: x/projects/README.md (hunk 4 above)
Comment (oren-lava) [reply x/projects/README.md]: done. see https://github.com/lavanet/lava/commit/576717fd44c05504a0127dc80eb39d09a153b012

File: x/projects/README.md (hunk 5 above)
Comment (oren-lava) [reply x/projects/README.md]: done. see https://github.com/lavanet/lava/commit/576717fd44c05504a0127dc80eb39d09a153b012

File: x/plans/README.md (hunk 6 above)
Comment (oren-lava) [reply x/plans/README.md]: done. see https://github.com/lavanet/lava/commit/8716c13495e79fb3c836254917cb6de2553db485

**Review by Yaroms — CHANGES_REQUESTED** [review]
can you also make it clear throughout the doc that changes in the project takes place in the next epoc?
developer keys, policies etc...

File: x/projects/README.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+```
+enum Kinds {
+    ADMIN = 1            // admin key
+    DEVELOPER = 2        // developer key
+    ADMIN+DEVELOPER = 3  // admin+developer key
+}
+```
+
+The admin key lets the admin edit the project's admin policy, enable/disable the project and add/delete other project keys. A single user can be admin in multiple projects. The subscription owner is automatically considered an admin in all of the project created under it.
```
Comment (shleikes) [inline x/projects/README.md:73]:
```suggestion
The admin key lets the admin edit the project's admin policy, enable/disable the project and add/delete other project keys. A single user can be admin in multiple projects. The subscription owner is automatically considered an admin in all of the projects created under it.
```

File: x/projects/README.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+The developer key lets a user to use the project, i.e., make RPC calls and use the project's CU. A single user (consumer) can only be a developer in one project. The project's developer can be viewed as a dapp developer that wants to use Lava to get RPC data for their dapp. To allow end users use their project to get RPC data, the dapp developer will need to use badges. For more details on badges, see [below](#badges).
+
+Note that the admin cannot use the project's CU like a developer, they can only edit the project's properties.
+
+### Badges
+
+Badges are a method to grant compute units (CU) from projects to other users. Given a subscription with projects, a developer key (in a project) can be used to generate a badge key, to be signed by an external badge-server. The badge is ephemeral, limited in time and in CU capacity; It can be used to get pairing with providers, and for requests from providers - which will be charged to the respective project after verification.
```
Comment (shleikes) [inline x/projects/README.md:81]: Can you rephrase the last sentence here, it is not very clear.
(Starting from "It can be used to get...")

File: x/projects/README.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+	CuAllocation  uint64  // badge CU limit
+	Epoch         uint64  // badge epoch
+	Address       string  // badge user address
+	LavaChainId   string  // lava chain ID (testnet/mainnet)
+	ProjectSig    []byte  // developer key signature (for verification)
+	VirtualEpoch  uint64  // used for emergency mode
+}
+```
+
+The badge's `Epoch` field is the epoch in which the badge is valid. When the epoch changes, the badge becomes invalid. The badge's `VirtualEpoch` is used in emergency mode (see `Downtime` module's README for more details).
```
Comment (shleikes) [inline x/projects/README.md:96]:
```suggestion
The badge's `Epoch` field is the epoch in which the badge is valid in. When the epoch changes, the badge becomes invalid. The badge's `VirtualEpoch` is used in emergency mode (see `Downtime` module's README for more details).
```

File: x/projects/README.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+        - collection:
+            api_interface: "tendermintrpc"
+            type: ""
+          extensions:
+            - "archive"
+          mixed: true
+    - chain_id: "*" # allows all other chains without specifying
+```
+
+All fields are not mandatory. An unfilled field will be replaced with a default value or be ignored.
```
Comment (shleikes) [inline x/projects/README.md:188]:
```suggestion
All fields are optional. An unfilled field will be replaced with a default value or be ignored.
```

File: x/projects/keeper/project.go [hunk 11]
```diff
@@ -326,6 +327,20 @@ func (k Keeper) SetProjectPolicy(ctx sdk.Context, projectIDs []string, policy *p
...
+	details := map[string]string{
+		"creator":     key,
+		"project_ids": projectIDsStr,
+		"policy":      policy.String(),
+	}
+
+	if setPolicyEnum == types.SET_ADMIN_POLICY {
+		utils.LogLavaEvent(ctx, k.Logger(ctx), types.SetAdminPolicyEventName, details, "set admin policy successfully")
+	} else if setPolicyEnum == types.SET_SUBSCRIPTION_POLICY {
+		utils.LogLavaEvent(ctx, k.Logger(ctx), types.SetSubscriptionPolicyEventName, details, "set admin policy successfully")
```
Comment (shleikes) [inline x/projects/keeper/project.go:343]: Is it supposed to be:
```suggestion
		utils.LogLavaEvent(ctx, k.Logger(ctx), types.SetSubscriptionPolicyEventName, details, "set subscription policy successfully")
```

**Review by shleikes — CHANGES_REQUESTED** [review]

File: x/projects/keeper/project.go (hunk 11 above)
Comment (oren-lava) [reply x/projects/keeper/project.go]: yes. nice catch. see here: https://github.com/lavanet/lava/commit/189b8756e705e6ccd58fd7e92b7231078bd5e5e0

File: x/projects/README.md (hunk 9 above)
Comment (oren-lava) [reply x/projects/README.md]: done. see https://github.com/lavanet/lava/commit/189b8756e705e6ccd58fd7e92b7231078bd5e5e0

File: x/projects/README.md (hunk 8 above)
Comment (oren-lava) [reply x/projects/README.md]: hope it's clearer now. see https://github.com/lavanet/lava/commit/189b8756e705e6ccd58fd7e92b7231078bd5e5e0

File: x/projects/README.md (hunk 7 above)
Comment (oren-lava) [reply x/projects/README.md]: done. see https://github.com/lavanet/lava/commit/189b8756e705e6ccd58fd7e92b7231078bd5e5e0

File: x/projects/README.md (hunk 8 above)
Comment (shleikes) [reply x/projects/README.md]: Perfect!

**Review by shleikes — APPROVED** [review]

File: x/projects/README.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+}
+```
+
+The project index is: `<subscription_owner_address>-<project_name>` (for example: `lava@1p7ezdua4ma90858h7zy9727wmqy53vk0mkdlvg-myProject`). When a consumer buys a subscription, an admin project is automatically generated with the index `<subscription_owner_address>-admin`.
+
+The project is subject to three types of policies: the plan's policy, the subscription's policy, and the admin policy. The plan's policy cannot be changed using transactions (only via a government proposal), while the other two policies can be changed using transactions. The effective policy that the project is subject to is the strictest policy calculated from all three policies. To query for the project's effective policy, use the `Pairing` module's query: `effective-policy` (see [below](#queries)).
+
+Projects are saved in a fixation store (see `FixationStore` module for reference). By using the snapshot ID, we can differentiate between different versions of the project.
+
+For more details regarding the `Policy` struct and its limitations, see [here](https://github.com/lavanet/lava/blob/main/x/plans/README.md#policy). For more details about `ProjectKeys`, see [below](#project-keys).
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/README.md:50]: done. see https://github.com/lavanet/lava/commit/be20a2ae9b26ea1430e798d71e215cd816e9aa66

[the remark this answers is not in the document]
Comment (Yaroms) [reply x/projects/README.md:50]: this is wrong
subscription policy is not the same policy for all the projects.
- subscription policy -  a policy defined by the subscription owner
- admin policy - a policy defined by the admins of the projects

[the remark this answers is not in the document]
Comment (oren-lava) [reply x/projects/README.md:50]: you're right. see https://github.com/lavanet/lava/commit/cbf5680a061780bc0aea2f21ea184bf857a2e542
